util update

Tharki-God opened the pull request:
added support for nested settings and fixed type for findInReactTree

this pr will close https://github.com/replugged-org/replugged/issues/383


sorry closed the old pr by mistake coz of renaming the branch. sorry

asportnoy reviewed (changes requested):
To prevent breaking changes, add a fallback as the original string with the dot included.

Tharki-God commented:
> To prevent breaking changes, add a fallback as the original string with the dot included.

i dont get what you mean. if you think not including . will break for not nested settings then its not the case ig

asportnoy commented:
If an addon is already relying on a key like `hello.world`, that data essentially will be wiped after this PR. So the initial behavior needs to be added as a fallback on the getter.

Tharki-God commented:
How do I add that? Any ideas?

asportnoy commented:
Just add a fallback to use the initial behavior if the new behavior is undefined

Tharki-God commented:
How do I decide if new behavior is undefined or not. What is the new behavior anyways

Tharki-God commented:
fixed types, and fallback value usage. also added a check if the key given is already in settings ignoring the path separator it will be treated as a key. 
well should be enough for what you wanted

asportnoy commented on src/renderer/util.ts:
```diff
@@ -183,29 +184,42 @@ type ValType<T> =
-    ? T[K] | undefined
-    : NonNullable<T[K]> | F;
-  onChange: (newValue: ValType<T[K]>) => void;
+  value: NonNullable<NestedType<T, P>>;
+  onChange: (newValue: ValType<NonNullable<NestedType<T, P>>>) => void;
```
You can't just wrap the whole thing in `NonNullable`, it needs to match how it was done before.

asportnoy commented on src/renderer/util.ts:
```diff
@@ -183,29 +184,42 @@ type ValType<T> =
 } {
-  const initial = settings.get(key, fallback);
-  const [value, setValue] = React.useState(initial);
+  const [initialKey, ...pathArray] = Object.keys(settings.all()).includes(key)
+    ? ([key] as [K])
+    : (key.split(".") as [K, ...string[]]);
+  const path = pathArray.join(".");
+  const initial = settings.get(initialKey, path.length ? ({} as T[K]) : (fallback as T[K]));
+  const [value, setValue] = React.useState<NonNullable<NestedType<T, P>>>(
+    path.length
+      ? (lodash.get(initial, path, fallback) as NonNullable<NestedType<T, P>>)
+      : (initial as NonNullable<NestedType<T, P>>),
+  );
```
This is all way more complicated than it needs to be. Just do something like `settings.get(key) ?? lodash.get(settings.all(), key) ?? defaultValue)`.

asportnoy commented on src/renderer/util.ts:
```diff
@@ -215,31 +229,28 @@ export function useSetting<
-    : NonNullable<T[K]> | F,
-  (newValue: ValType<T[K]>) => void,
-] {
+): [NonNullable<NestedType<T, P>>, (newValue: ValType<NonNullable<NestedType<T, P>>>) => void] {
```
Same type issue as above

asportnoy commented on src/renderer/util.ts:
```diff
@@ -361,11 +372,11 @@ export function findInTree(
 export function findInReactTree(
-  tree: Tree,
+  tree: Tree | ReactElement,
   searchFilter: TreeFilter,
   maxRecursion = 100,
-): Tree | null | undefined {
-  return findInTree(tree, searchFilter, {
+): Tree | ReactElement | null | undefined {
+  return findInTree(tree as Tree, searchFilter, {
     walkable: ["props", "children", "child", "sibling"],
     maxRecursion,
   });
```
Unrelated change, does not belong in this PR

lexisother commented on bin.mjs:
Can you remove this?

lexisother commented on src/renderer/util.ts:
```diff
@@ -85,7 +85,6 @@ export function forceUpdateElement(selector: string, all = false): void {
     all ? [...document.querySelectorAll(selector)] : [document.querySelector(selector)]
   ).filter(Boolean) as Element[];
 
-  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- May not actually have forceUpdate
```
Why was this removed? Now there's an ESLint warning.

YofukashiNo replied:
by mistake ig, i forgor

lexisother commented on src/renderer/util.ts:
```diff
@@ -172,41 +171,56 @@ export async function goToOrJoinServer(invite: string): Promise<void> {
     throw new Error("Could not find openExternal");
   }
-  await mod(url);
+  return await mod(url);
```
Why did you add this? `mod` returns `void`, adding `return` here is pointless.

fedeericodl reviewed (approved):
I've added type safety for nested settings.
LGTM

lexisother commented:
Could you give some usage examples? e.g. with `src/types/settings.ts`, how would it look like if `autoApplyQuickCss` became `quickCss.autoApply` and `addonEmbeds` became `qol.addonEmbeds`?

I understand the `useSetting` call would become `useSetting(generalSettings, 'qol.addonEmbeds')` but what about that interface and the `defaultSettings` object? Would you do it like this?
```ts
export type GeneralSettings = {
  'quickCss.autoApply'?: boolean;
  'qol.addonEmbeds'?: boolean;
};

export const defaultSettings = {
  'quickCss.autoApply': false,
  'qol.addonEmbeds': true,
} satisfies Partial<GeneralSettings>;
```

lexisother commented:
Implementing the example I've mentioned just now yields the following result. Why is `qol` not being put into an object?

<img width="276" height="146" alt="image" src="https://github.com/user-attachments/assets/7d218663-0203-40c0-9efa-0b867eec1559" />

yofukashino commented:
> Could you give some usage examples? e.g. with `src/types/settings.ts`, how would it look like if `autoApplyQuickCss` became `quickCss.autoApply` and `addonEmbeds` became `qol.addonEmbeds`?
> 
> I understand the `useSetting` call would become `useSetting(generalSettings, 'qol.addonEmbeds')` but what about that interface and the `defaultSettings` object? Would you do it like this?
> 
> ```ts
> export type GeneralSettings = {
>   'quickCss.autoApply'?: boolean;
>   'qol.addonEmbeds'?: boolean;
> };
> 
> export const defaultSettings = {
>   'quickCss.autoApply': false,
>   'qol.addonEmbeds': true,
> } satisfies Partial<GeneralSettings>;
> ```

Probably like a normal object. 

Example from show hidden channels

```ts
  export interface Settings {
    hiddenChannelIcon: string;
    customHiddenChannelIcon: string;
    faded: boolean;
    sort: string;
    showPerms: boolean;
    showAdmin: string;
    stopMarkingUnread: boolean;
    shouldShowEmptyCategory: boolean;
    alwaysCollapse: boolean;
    debugMode: boolean;
    channels: {
      GUILD_TEXT: boolean;
      GUILD_VOICE: boolean;
      GUILD_ANNOUNCEMENT: boolean;
      GUILD_STORE: boolean;
      GUILD_STAGE_VOICE: boolean;
      GUILD_FORUM: boolean;
    };
    blacklistedGuilds: Record<string, boolean>;
    collapsed: Record<string, boolean>;
  }
}
    ```
    
    Used like 
    ```ts
      {Object.keys(SettingValues.get("channels", defaultSettings.channels)).map(
          (type: keyof typeof defaultSettings.channels) => (
            <SwitchItem
              key={type}
              {...Utils.useSetting(
                SettingValues,
                `channels.${type}`,
                defaultSettings.channels[type],
              )}>
              {`Show ${Utils.capitalizeFirst(type.split("_")[1])}${
                type.split("_").length === 3 ? ` ${Utils.capitalizeFirst(type.split("_")[2])}` : ""
              } Channels`}
            </SwitchItem>
          ),
        )}```

yofukashino commented:
> Mind you that `generalSettings.get('qol').addonEmbeds` _does_ work. That's fine in my book, but I'm wondering if you want to adjust the implementation there too.


This is just meant for hooks as reading like this aint that big of mess but the custom hooks for this are

lexisother commented:
Would you be opposed to adapting the `SettingsManager` as well? If you're fetching a single setting it'd be nice to just give it `qol.addonEmbeds` instead of `.get('qol').addonEmbeds`.

yofukashino commented:
> Would you be opposed to adapting the `SettingsManager` as well? If you're fetching a single setting it'd be nice to just give it `qol.addonEmbeds` as an argument instead of having to do `.get('qol').addonEmbeds`.

I wouldn't be opposed, but I don't think it's required as you dont need to write something extra just the position.

lexisother commented:
I would prefer it be added in this PR. I'm of the opinion that if `useSetting` supports it then so should the rest of the API.

yofukashino commented:
> I would prefer it be added in this PR. I'm of the opinion that if `useSetting` supports it then so should the rest of the API.

Same for set?  and has? imo its required only in the hook
